### PR TITLE
Move uploading media to WP Media Library to `MediaImportService`

### PR DESCRIPTION
**Pull request by crazytonyli**
There is the usual accessing managed object in a wrong thread issue (I'll highlight a few in the PR comments) in the implementation of uploading a media asset to the WP Media Library.

This PR rewrites (large chunk of the original code is still reused) the uploading logic to make the Core Data operations safer, by adopting the new writer API.

To help with the review, I'll quickly described what happens during uploading a media asset to WP media library. The entry point is [the `MediaCoordinator.addMedia` function](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaCoordinator.swift#L161), which can be called during editing a post or simply uploading an photo (on device or from the free online photo library) on the "Media" screen. This function first imports the photo into the app's sandbox, using [`MediaImportService`](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaImportService.swift#L49). Then it generates a thumbnail using [`MediaThumbnailService`](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaThumbnailService.swift#L47). At the same time, it starts to upload the imported media to the WP media library, using [`MediaService`](https://github.com/wordpress-mobile/WordPress-iOS/blob/8f011215fe468af2417c70acdf7d4daa583571ae/WordPress/Classes/Services/MediaService.m#L199).

The refactor does not change the above logic, but moves some code around, makes the Core Data operations safer, and changed a few call sites to adopt to the new API.

## Test instructions

Verify all the places that can upload a photo to WP media library. I've tested uploading a new site icon, uploading a photo to the Media Library on "Media" screen, uploading a photo to a post in the post editor.

## Regression Notes

1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
What described in the "Test instructions"

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Services/MediaImportService.swift
Comment (crazytonyli) [inline WordPress/Classes/Services/MediaImportService.swift:59]: ⚠️ Potential issue: `media` is changed in the `managedObjectContext` queue, but there is no guarantee it's bound to the context object.

Comment (crazytonyli) [reply WordPress/Classes/Services/MediaImportService.swift:59]: ⚠️ Potential issue: `media` is send to the `onCompletion` closure on the main thread, but there is no guarantee that `media` is bound to the main context.

File: WordPress/Classes/Services/MediaService.m
Comment (crazytonyli) [inline WordPress/Classes/Services/MediaService.m:84]: ⚠️ Potential issue: `media` is modified in the main thread since this `completionWithMedia` is [called from the main thread](https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/files#r1136467532), but there is no guarantee `media` is bound to the main context.

File: WordPress/Classes/Services/MediaService.m
Comment (crazytonyli) [inline WordPress/Classes/Services/MediaService.m:38]: ⚠️ Potential issue: this functions returns a `Media` instance with no guarantee what context object this instance is bound to.

File: WordPress/Classes/Services/MediaService.m
Comment (crazytonyli) [inline WordPress/Classes/Services/MediaService.m:119]: This function is moved to `MediaImportService` and duplicated as two different variants: one that returns a `Media` instance which can only be called from the main thread, and the other that can be called from any thread.

File: WordPress/Classes/Services/MediaImportService.swift
Comment (crazytonyli) [inline WordPress/Classes/Services/MediaImportService.swift:59]: This API doc explains why there are two `createMedia` functions in this class.

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (crazytonyli) [inline WordPress/Classes/Services/MediaCoordinator.swift:365]: I think this change—using the main context—fixes [the issue mentioned in this comment](https://github.com/wordpress-mobile/WordPress-iOS/issues/20298#issuecomment-1465319707).

**Comment by mokagio** [comment]
Just a note to say that I saw this PR, but haven't gotten an open chunk of time long enough to look at it yet.

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaCoordinator.swift]: Nitpick. Does the coordinator need to be stored this way? It's an `NSObject` after all so passed by reference.

https://github.com/wordpress-mobile/WordPress-iOS/blob/3dfbb57c56a07b1bd224656f7f36398e78b2316a/WordPress/Classes/Services/MediaCoordinator.swift#L33-L38

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaCoordinator.swift]: This one, too, could be shorter

```swift
addMedia(from: asset, post: post, coordinator: coordinator(for: post), analyticsInfo: analyticsInfo)
```

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaCoordinator.swift]: Noticed that `self` is not captured weakly here, unlike the other callbacks. I think it's fine in that `self` should always be available when `receiveUpdate` is called, but just checking 😄

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaCoordinator.swift:217]: As soon as I saw that `media, error` pair in the code about my mind screamed: Use `Result`! But here I see that it wouldn't work properly in we can have both an error and a media 😞

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (mokagio) [reply WordPress/Classes/Services/MediaCoordinator.swift:365]: In such case, what do you think of adding a link to the issue?

```suggestion
        // 
        // See https://github.com/wordpress-mobile/WordPress-iOS/issues/20298
        let service = self.mediaServiceFactory.create(coreDataStack.mainContext)
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    /// Create a media object using the URL provided as the source of media.
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    /// This functions returns a `Media` instance. To ensure the returned `Media` instance continues to be a valid
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [reply WordPress/Classes/Services/MediaImportService.swift:59]: Informative but not verbose 👍

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]: Nitpick:

```suggestion
    ///   - exportable: an object that conforms to `ExportableAsset`
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    ///   - thumbnailCallback: a block that will be invoked when the thumbnail for the media object is ready
```

Comment (mokagio) [reply WordPress/Classes/Services/MediaImportService.swift]: Also, Swift nitpick:

```suggestion
    ///   - thumbnailCallback: a closure that will be invoked when the thumbail for the media object is ready
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    ///   - completion: a closure that will be invoked when the media is created, on success it will return a valid Media
    ///         object, on failure it will return a nil Media and an error object with the details.
```

Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    ///   - completion: a block that will be invoked when the media is created, on success it will return a valid `Media`
    ///         object, on failure it will return a `nil` `Media` and an error object with the details.
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]: 🤔 this is not in line with the return value, though?

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift:79]: Yes, might as well crash on our own terms here.

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    ///   - post: the optional post object to associate to the media
```

I'm guessing the post is optional because we can upload assets via the media library itself, in which case they wouldn't be used in any post?

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift:131]: The code never has both `Media` and `Error` `nil` at the same time. However, I'm guessing this doesn't use `Result` because of `@objc`?

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]: Nitpick: Just for consistency

```suggestion
    private func `import`(
        _ exportable: ExportableAsset, 
        to media: Media,
        blog: Blog,
        thumbnailCallback: ((Media, URL) -> Void)?,
        completion: @escaping (Result<Media, Error>) -> Void
    ) -> Progress {
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift:188]: I wonder why this GitHub link was shortened with Git.io 🤔 

To keep the comment compact? 🤷‍♂️f

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    ///         `absoluteLocalURL`. This closure is called on the main thread. The closure's `media` argument is also
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    ///     - onError: Called if an error was encountered during creation, error convertible to `NSError` with a
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift:274]: Have you considered using "allowed"?

**Update:** Oh, I see that name was already used. 🤔 No point in using a different one, then, it would end up being more confusing than useful, even if we were to agree that "allowed" is in itself clearer than "allowable".

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    /// Generate a thumbnail image for the Media asset so that consumers of the `absoluteThumbnailLocalURL` property
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    /// will have an image ready to load, without using the async methods provided via `MediaThumbnailService`.
```

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift]:
```suggestion
    /// Note: Ideally we wouldn't need this at all, but the synchronous usage of `absoluteThumbnailLocalURL` across the code-base
    ///       to load a thumbnail image is relied on quite heavily. In the future, transitioning to asynchronous thumbnail loading
    ///       via the new thumbnail service methods is much preferred, but would indeed take a good bit of refactoring away from
    ///       using `absoluteThumbnailLocalURL`.
```

Comment (mokagio) [reply WordPress/Classes/Services/MediaImportService.swift]: Also, 👍 looking forward to that.

File: WordPress/Classes/Services/MediaImportService.swift
Comment (mokagio) [inline WordPress/Classes/Services/MediaImportService.swift:365]: Nitpick. This doesn't access anything from `self`. Could it become a constant so we don't need to compute it on every access?

**Review by mokagio — COMMENTED** [review]
Most of my comments are nitpicks.

I'm running short on time ATM, so I haven't tested this locally.

My only concern is that with test coverage. This is a refactor, so it should mostly rely on previous tests to ensure that the desired behavior hasn't been affected. But, how comprehensive and trustworthy are those tests?

For example, I modified the code randomly like this and got no test failures 😢 

```diff
--- a/WordPress/Classes/Services/MediaImportService.swift
+++ b/WordPress/Classes/Services/MediaImportService.swift
@@ -136,29 +136,30 @@ class MediaImportService: NSObject {
             try context.obtainPermanentIDs(for: [media])
             return media.objectID
         }, completion: { (result: Result<NSManagedObjectID, Error>) in
-            let transformed = result.flatMap { mediaObjectID in
-                Result {
-                    (
-                        try self.coreDataStack.mainContext.existingObject(with: mediaObjectID) as! Media,
-                        try self.coreDataStack.mainContext.existingObject(with: blog.objectID) as! Blog
-                    )
-                }
-            }
-            switch transformed {
-            case let .success((media, blog)):
-                let progress = self.import(exportable, to: media, blog: blog, thumbnailCallback: thumbnailCallback) {
-                    switch $0 {
-                    case let .success(media):
-                        completion(media, nil)
-                    case let .failure(error):
-                        completion(nil, error)
-                    }
-                }
-                createProgress.addChild(progress, withPendingUnitCount: 1)
-                receiveUpdate?(media)
-            case let .failure(error):
-                completion(nil, error)
-            }
+//            let transformed = result.flatMap { mediaObjectID in
+//                Result {
+//                    (
+//                        try self.coreDataStack.mainContext.existingObject(with: mediaObjectID) as! Media,
+//                        try self.coreDataStack.mainContext.existingObject(with: blog.objectID) as! Blog
+//                    )
+//                }
+//            }
+//            switch transformed {
+//            case let .success((media, blog)):
+//                let progress = self.import(exportable, to: media, blog: blog, thumbnailCallback: thumbnailCallback) {
+//                    switch $0 {
+//                    case let .success(media):
+//                        completion(media, nil)
+//                    case let .failure(error):
+//                        completion(nil, error)
+//                    }
+//                }
+//                createProgress.addChild(progress, withPendingUnitCount: 1)
+//                receiveUpdate?(media)
+//            case let .failure(error):
+//                completion(nil, error)
+//            }
+            completion(nil, nil)
         }, on: .main)
         return createProgress
     }
```

**Comment by mokagio** [comment]
I'll move the milestone to 22.1 because I'm about to start the 22.0 code freeze and this hasn't been approved yet. Given this is a refactor/rewrite, we should be able to afford not shipping to beta in this release cycle. Feel free to revert if necessary.

**Comment by crazytonyli** [comment]
> My only concern is that with test coverage. This is a refactor, so it should mostly rely on previous tests to ensure that the desired behavior hasn't been affected. But, how comprehensive and trustworthy are those tests?

I wouldn't trust the unit tests to assure any quality on this refactor, unfortunately.

Incorrectly accessing the created `Media` instance in a wrong thread—the issue this PR tries to address—doesn't cause any obvious issues to the app. [This uploading media issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/20322#discussion_r1136479772) might be one of them, but it's not easily reproducible. Thus, it's hard to verify that the refactor indeed achieve what it sets out to achieve.

A more valuable test coverage would be making sure this refactor doesn't break things, uploading Media to the WP Media Library in particular. IMO, for now, the most effective way to ensure the coverage is testing all these features manually. But I can spend sometime looking into adding some UI tests, which will benefit us in long run, as I can see the Media area sure needs some further refactors similar to this PR.

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (crazytonyli) [reply WordPress/Classes/Services/MediaCoordinator.swift]: Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/65fd4e4396d0a6d87dd1500a79a2d00d31cb4f8b

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (crazytonyli) [reply WordPress/Classes/Services/MediaCoordinator.swift]: Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/cce100461b3f39858453123eff08ec913b0ad03b

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (crazytonyli) [reply WordPress/Classes/Services/MediaCoordinator.swift]: Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/4d4fd8ab941f9f850e961cf679e53b74177372fc — it's good to keep them consistent.

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (crazytonyli) [reply WordPress/Classes/Services/MediaCoordinator.swift:365]: Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/65b7249674e554113fb4e6b983d8ff4997660d02

File: WordPress/Classes/Services/MediaCoordinator.swift
Comment (crazytonyli) [reply WordPress/Classes/Services/MediaCoordinator.swift:217]: Good catch! I now noticed there are a couple places I should have passed the `Media` instance along with the error, which are updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/e8b02b8875a61da11f78580618339e50aace2cf2.

**Review by mokagio — APPROVED** [review]
Thank you for considering my suggestion.

Regarding test coverage and trustworthiness. I see what you mean with the PR [tackling the thread access issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/20322#issuecomment-1478805436) 🤔 That would be indeed hard to unit test.

I remain worried about the ratio between the amount of changes that went into this and the unit tests covering them. But I don't want this to come across as a critique to your work or a blocker, just an observation and maybe a _collective_ call to action to add _some_ unit tests.

>  But I can spend sometime looking into adding some UI tests, which will benefit us in long run, as I can see the Media area sure needs some further refactors similar to this PR.

I'd love to see that happening. However I do wonder how feasible it would be, given that we rely on API mocks? I presume this would need to actually upload the images, or would we be find with stubbed responses? 🤔 

Also, before investing in that, I'd recommend checking if there are plans to change the UX soon, just to avoid wasted effort.

**Comment by crazytonyli** [comment]
> I remain worried about the ratio between the amount of changes

I agree. I am worried too. 🥲 

I wanted to break the changes down to small PRs. I think I'll have another go, not just for the sake of PR review, for myself too, so that I have more confident that the changes don't break things.

> But I don't want this to come across as a critique to your work or a blocker

I didn't take it that way. 😄 Thanks for your sharp eyes, that I caught [a changed behavior](https://github.com/wordpress-mobile/WordPress-iOS/pull/20322#discussion_r1144151845), which I'm not sure if will cause issues, but still, it's a change that I didn't intend to make.

**Comment by mokagio** [comment]
> I wanted to break the changes down to small PRs. I think I'll have another go, not just for the sake of PR review, for myself too, so that I have more confident that the changes don't break things.

Sounds good to me. We have this approved and we can keep it warm, ready to merge if the need arises.

**Comment by crazytonyli** [comment]
@mokagio I had made an attempt to split this diff, but couldn't get it right, they are all very much tangled together. I've [added a release note](https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/f6cc6e4980da79a98e2064f2544ed5406c9b31c1) highlighting this is a change that may have a high impact on the app. Since I've got your approval, I'll merge this PR and see if there is any issue from the beta testing phase.

File: RELEASE-NOTES.txt
Comment (mokagio) [inline RELEASE-NOTES.txt]:
```suggestion
* [***] [internal] Refactor uploading photos (from the device photo, the Free Photo library, and other sources) to the WordPress Media Library. Affected areas are where you can choose a photo and upload, including the "Media" screen, adding images to a post, updating site icon, etc. [#20322]
```

**Comment by mokagio** [comment]
@crazytonyli no worries. Auto-merge enabled 🤖

**Pull request merged**
